com.valvesoftware.SteamLink.yml

```diff
@@ -10,6 +10,7 @@ finish_args:
   - --env=SDL_GAMECONTROLLERCONFIG_FILE=/var/data/Valve Corporation/SteamLink/controller_map.txt
   - --share=ipc
   - --share=network
+  - --socket=fallback-x11
   - --socket=pulseaudio
   - --socket=wayland
   - --socket=x11
```

Comment:
Shouldn't the legacy `x11` socket be removed if `fallback-x11` is added?

Reply:
Hmm, it seems that some apps only have the fallback-x11 socket only while others have both x11 and fallback-x11. I am not entirely sure what is correct.

Reply:
Having both `x11` and `fallback-x11` is more compatible with Flatpak versions before 1.0.

In Flatpak 1.0+, `flatpak-x11` means: try `wayland`. If Wayland is possible, behave as though `x11` was *not* allowed. If Wayland *isn't* possible, behave as though `x11` *was* allowed. It's irrelevant whether `x11` is actually in the permissions or not.

In Flatpak older than 1.0, `flatpak-x11` is unsupported. We want these versions to see both `--socket=x11` and `--socket=wayland`, so that the app can use X11 on non-Wayland-capable systems (which is more permissive, but Flatpak older than 1.0 is no longer supported and has various known flaws in its sandboxing anyway, so if you're still using one of those versions, giving out X11 access is presumably the least of your concerns).

Reply:
Ah! Thanks for the explanation. :-)

Reply:
that's true, but flatpak-builder automatically adds `--socket=x11` if `--socket=fallback-x11` is specified (See: https://github.com/flatpak/flatpak-builder/blob/c765f46feb6a0ad2096084d2f470199d6fdb454f/src/builder-flatpak-utils.c#L1805-L1806)
So effectively there is no difference between only specifying `--socket=fallback-x11` or having both.

Reply:
Flatpak builder here is irrelevant.

The relevant code is in flatpak.

https://github.com/flatpak/flatpak/blob/38eac0729318cae02b05d7f5673628028d075a58/common/flatpak-run.c#L1606

There is a difference:  with fallback x11 it only add the x11 socket if there is no wayland.

Reply:
> Hmm, it seems that some apps only have the fallback-x11 socket while others have both x11 and fallback-x11. I am not entirely sure what is correct.

`fallback-x11` and `x11` in the same manifest ought to be considered as a bug. `wayland` and `x11` in the same manifest is likely an oversight (or legacy, ie never changed), albeit I have seen a few case where it's required.

Reply:
> `flatpak-builder` here is irrelevant.
> 
since this flatpak gets build using `flatpak-builder` it's not irrelevant!

for example take `org.gnome.Epiphany` a flatpak that only has `--socket=fallback-x11` in it's [flatpak-builder manifest](https://github.com/flathub/org.gnome.Epiphany/blob/master/org.gnome.Epiphany.json#L13)

If you install it and take a look at the permissions is has `x11` and `fallback-x11` since flatpak-builder added `x11` implicitly.
```
> flatpak info -M org.gnome.Epiphany
[Context]
shared=network;ipc;
sockets=x11;wayland;pulseaudio;fallback-x11;
devices=dri;
filesystems=xdg-download;/run/.heim_org.h5l.kcm-socket;

[System Bus Policy]
org.freedesktop.GeoClue2=talk
```